### PR TITLE
BRING-17: Implement ComponentAnnotationScanner

### PR DESCRIPTION
**Pull request by withgodinheart**
# Why:
It allows us to find the beans to put in the ApplicationContext

# How:
1. Scan provided package
2. Find all classes annotated with **Component** annotation
3. Create **BeanDefinition** based on these classes
4. If the class has fields annotated with **Autowired** annotation, then add dependencies to **BeanDefinition#dependsOnBeans**
5. Return **Collection** of **BeanDefinition**

# Testing:
Test data is taken from **org.blyznytsia.scanner.data** test package

Screenshot 2022-07-21 at 1 50 43 PM

### CODE REVIEW

File: src/main/java/org/blyznytsia/scanner/ComponentAnnotationScanner.java [hunk 1]
```diff
@@ -1,16 +1,90 @@
...
+
+    /**
+     * @param targetClass object of type {@link Class} that is annotated
+     *                    with @{@link Component} annotation
+     * @return {@link java.util.Collection} of {@link String} that represents bean dependencies
+     */
+    private List<String> getAutowiredFields(final Class<?> targetClass) {
+        return Arrays.stream(targetClass.getDeclaredFields())
+                .filter(el -> el.isAnnotationPresent(AUTOWIRED_ANNOTATION))
+                .filter(el -> el.getType().isAnnotationPresent(COMPONENT_ANNOTATION))
```
Comment (withgodinheart) [inline src/main/java/org/blyznytsia/scanner/ComponentAnnotationScanner.java:86]: I think I should delete this line

File: src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        Assertions.assertThat(checkFullPresence(beanDefinitions, SERVICE2_NAME, SERVICE1_NAME)).isTrue();
+
+        final var service1 = getBeanDefinitionByName(beanDefinitions, SERVICE1_NAME);
+        final var service2 = getBeanDefinitionByName(beanDefinitions, SERVICE2_NAME);
+
+        Assertions.assertThat(beanDefinitions).hasSize(BEAN_DEFINITIONS_SIZE);
+        Assertions.assertThat(service1.getDependsOnBeans()).isEmpty();
+        Assertions.assertThat(service2.getDependsOnBeans()).hasSize(SERVICE2_DEPENDENCIES_SIZE);
+        Assertions.assertThat(service2.getDependsOnBeans().get(0)).isEqualTo(SERVICE2_DEPENDENCY_NAME);
+    }
```
Comment (kebann) [inline src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java:35]: can be refactored to be as follows to take advantage of AssertJ API

``` java
        var beanDefinitions =
                new ComponentAnnotationScanner().scan(PACKAGE_NAME);

        assertThat(beanDefinitions)
                .map(BeanDefinition::getName)
                // the size check is included
                .containsOnly(SERVICE2_NAME, SERVICE1_NAME);

        var service1 = getBeanDefinitionByName(beanDefinitions, SERVICE1_NAME);
        var service2 = getBeanDefinitionByName(beanDefinitions, SERVICE2_NAME);

        assertThat(service1.getDependsOnBeans()).isEmpty();
        assertThat(service2.getDependsOnBeans())
                .hasSize(SERVICE2_DEPENDENCIES_SIZE)
                .first()
                .isEqualTo(SERVICE2_DEPENDENCY_NAME);
```

File: src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+    private static final String PACKAGE_NAME = "org.blyznytsia.scanner.data";
+    private static final String SERVICE1_NAME = "anotherNameService";
+    private static final String SERVICE2_NAME = "testService2";
+    private static final int BEAN_DEFINITIONS_SIZE = 2;
+    private static final int SERVICE2_DEPENDENCIES_SIZE = 1;
+    private static final String SERVICE2_DEPENDENCY_NAME = TestService1.class.getName();
+
+    @Test
+    void scan_givenValidPackage_shouldReturnValidBeanDefinitionsList() {
```
Comment (kebann) [inline src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java:22]: We can discuss the test naming convention on a daily call and stick to it to keep things consistent.

IMO, `shouldReturnValidBeanDefinitionsAfterScanning` should suffice and more details should be provided via `@DisplayName`

**Review by kebann — CHANGES_REQUESTED** [review]

File: src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java (hunk 3 above)
Comment (kebann) [reply src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java:22]: agreed

File: src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        final var beanDefinitions =
+                new ComponentAnnotationScanner().scan(PACKAGE_NAME);
+        beanDefinitions.forEach(System.out::println);
+        Assertions.assertThat(checkFullPresence(beanDefinitions, SERVICE2_NAME, SERVICE1_NAME)).isTrue();
+
+        final var service1 = getBeanDefinitionByName(beanDefinitions, SERVICE1_NAME);
```
Comment (PavloChechehov) [inline src/test/java/org/blyznytsia/scanner/ComponentAnnotationScannerTest.java:28]: I suppose we no need to overuse the `final` word, except in the case where it is vital

**Pull request merged**